src/printer.rs

```diff
@@ -378,63 +359,62 @@ impl<W: WriteColor> Printer<W> {
         }
     }
 
-    fn write_heading<P: AsRef<Path>>(&mut self, path: P) {
-        let _ = self.wtr.set_color(self.colors.path());
-        self.write_path(path.as_ref());
-        let _ = self.wtr.reset();
+    fn separator(&mut self, sep: &[u8]) {
+        self.write_colored(&sep, |colors| { colors.separator() });
+    }
+
+    fn write_path_sep(&mut self, sep: u8) {
         if self.null {
             self.write(b"\x00");
         } else {
-            self.write_eol();
+            self.separator(&[sep]);
         }
     }
 
-    fn write_non_heading_path<P: AsRef<Path>>(&mut self, path: P) {
-        let _ = self.wtr.set_color(self.colors.path());
-        self.write_path(path.as_ref());
-        let _ = self.wtr.reset();
+    fn write_path_eol(&mut self) {
         if self.null {
             self.write(b"\x00");
         } else {
-            self.write(b":");
+            self.write_eol();
+        }
+    }
+
+    fn write_path<P: AsRef<Path>>(&mut self, path: P) {
+        let path_str = self.path_to_str_native(&path);
+        match self.path_separator {
+            None => self.write_colored(path_str, |colors| { colors.path() }),
+            Some(sep) => {
+                let transformed_path: Vec<_> = path_str.iter().map(|&b| {
+                    if b == b'/' || (cfg!(windows) && b == b'\\') {
+                        sep
+                    } else {
+                        b
+                    }
+                }).collect();
```

Comment:
https://doc.rust-lang.org/std/primitive.slice.html#method.to_vec

The above `let mut path = path.to_vec();` was that allocation. I replaced for with map. No additional allocations are made by this code.